### PR TITLE
Fix incomplete backup handling

### LINKED ISSUE #133: POST_COMMAND_FAILURE triggering if backup completes
I only have a notification setup for POST_COMMAND_FAILURE, not the new _INCOMPLETE hook.

However I am seeing the following in my logs (and getting a notification);

```
Starting Backup at 2022-05-12 04:00:00
using parent snapshot 7df93e77
error: lstat /data/openhab/userdata/jsondb/backup/1652285002376--org.eclipse.smarthome.core.thing.Thing.json: no such file or directory
error: lstat /data/openhab/userdata/jsondb/backup/1652285006119--org.eclipse.smarthome.core.thing.Thing.json: no such file or directory
Files:         331 new,   243 changed, 45348 unmodified
Dirs:          184 new,  2651 changed, 18748 unmodified
Added to the repo: 1.323 GiB
processed 45922 files, 14.358 GiB in 5:38
snapshot d1fb0fda saved
Warning: at least one source file could not be read
curl -X POST --data "{\"title\": \"[restic-rest] Backup failed\", \"body\": \"Backup of docker swarm data volumes to local backup server (backup.home) failed\"}" http://notify:5000
```

It appears as tho the _FAILURE hook is being called for non-critical failures, i.e. missing file. The backup completes and the snapshot is created so I am not expecting to see any failure notification in this instance.

### PR DESCRIPTION
**Pull request by djmaze**
Allow configuring if an incomplete backup should be a success or not.

Adds `SUCCESS_ON_INCOMPLETE_BACKUP`, defaulting to `false`.

Should fix #133

### CODE REVIEW

**Comment by sumnerboy12** [comment]
Does this need a review?

**Comment by djmaze** [comment]
Yes and no. I tried to use this but it did not work as expected. It signaled incomplete backups as failed IIRC..

This kind of stuff is hard to test. Would be cool if someone else could test it as well.

File: backup [hunk 1]
```diff
@@ -54,10 +66,21 @@ restic --repo="${RESTIC_REPOSITORY}" backup "${backup_args[@]}" "${tag_options[@
 rc=$?
 set -e
 
+success="true"
 if [ $rc -ne 0 ]; then
-  if [ $rc -eq 3 ] && [ -n "${POST_COMMANDS_INCOMPLETE:-}" ]; then
-  	run_commands "${POST_COMMANDS_INCOMPLETE:-}"
+  if [ $rc -eq 3 ]; then
+    if [ "$(check_bool "${SUCCESS_ON_INCOMPLETE_BACKUP:-}")" = "true" ]; then
```
Comment (sumnerboy12) [inline backup:72]: Should this check be `== "true"`?

File: backup [hunk 2]
```diff
@@ -54,10 +66,21 @@ restic --repo="${RESTIC_REPOSITORY}" backup "${backup_args[@]}" "${tag_options[@
...
+      run_commands "${POST_COMMANDS_INCOMPLETE:-}"
+    else
+      success="false"
+      if [ -n "${POST_COMMANDS_INCOMPLETE:-}" ]; then
```
Comment (sumnerboy12) [inline backup:76]: Can this `if` statement be removed? You don't perform this check before calling `POST_COMMANDS_INCOMPLETE` above on line 73. I think (and I am not a bash expert) that `run_commands` checks the command exists before running anyway?

File: backup [hunk 3]
```diff
@@ -54,10 +66,21 @@ restic --repo="${RESTIC_REPOSITORY}" backup "${backup_args[@]}" "${tag_options[@
...
+        run_commands "${POST_COMMANDS_INCOMPLETE:-}"
```
Comment (sumnerboy12) [inline backup:77]: Shouldn't you be running `_INCOMPLETE` and `_FAILURE` commands in this scenario, always?

File: backup [hunk 4]
```diff
@@ -54,10 +66,21 @@ restic --repo="${RESTIC_REPOSITORY}" backup "${backup_args[@]}" "${tag_options[@
...
+      else
+        run_commands "${POST_COMMANDS_FAILURE:-}"
+      fi
+    fi
   else
+    success="false"
   	run_commands "${POST_COMMANDS_FAILURE:-}"
   fi
 fi
```
Comment (sumnerboy12) [inline backup:86]: Could this be simplified to something like;

```bash
success="true"
if [ $rc -ne 0 ]; then
  if [ $rc -eq 3 ]; then
    run_commands "${POST_COMMANDS_INCOMPLETE:-}"
    success="$(check_bool "${SUCCESS_ON_INCOMPLETE_BACKUP:-}")"
  else
    success="false"
  fi
fi

if [ "$success" == "true" ]; then
  echo Backup successful
else
  echo Backup failed
fi

if [ -n "${RESTIC_FORGET_ARGS}" ]; then
  IFS=" " read -r -a RESTIC_FORGET_ARGS <<< "$RESTIC_FORGET_ARGS"
  echo Forget about old snapshots based on RESTIC_FORGET_ARGS = "${RESTIC_FORGET_ARGS[@]}"
  restic forget "${tag_options[@]}" "${RESTIC_FORGET_ARGS[@]}"
fi

end=$(date +%s)
echo Finished backup at $(date +"%Y-%m-%d %H:%M:%S") after $((end-start)) seconds

if [ "$success" == "true" ]; then
  run_commands "${POST_COMMANDS_SUCCESS:-}"
else
  run_commands "${POST_COMMANDS_FAILURE:-}"
fi
```

**Review by sumnerboy12 — CHANGES_REQUESTED** [review]
Noticed a few things, which might have been why you were not seeing the expected outcomes?

File: backup (hunk 1 above)
Comment (djmaze) [reply backup:72]: No, [this is correct](https://stackoverflow.com/a/10849346) as we are using the old sh syntax with single brackets only.

File: backup (hunk 3 above)
Comment (djmaze) [reply backup:77]: No, the main point of this PR is to _not_ trigger the "fatal failure" commands if the backup is only incomplete.

But if the user did not set `POST_COMMANDS_INCOMPLETE`, the behaviour should be as before (running `POST_COMMANDS_FAILURE`), in order to be able to preserve the old behaviour and to be backwards compatible.

File: backup (hunk 2 above)
Comment (djmaze) [reply backup:76]: No, as we explicitely need the `else` branch. This is answered in detail below.

File: backup (hunk 4 above)
Comment (djmaze) [reply backup:80]: Thanks, that was a good inspiration for a refactoring which I now did in 2456f18ad4714cdaae68e14a16cd27c2e5bed193.

Comment (sumnerboy12) [reply backup:80]: Nice - is it passing your internal testing now? I am unable to test currently sorry.

Comment (djmaze) [reply backup:80]: The shellspec tests are still passing. Now need to test IRL.

**Review by sumnerboy12 — APPROVED** [review]

File: backup (hunk 4 above)
Comment (sumnerboy12) [reply backup:80]: Sorry to keep bugging you @djmaze - how has your IRL testing been going?

Comment (djmaze) [reply backup:80]: @sumnerboy12 Sorry for the late reply. I have connection problems on my test cluster which I need to fix first.

**Comment by djmaze** [comment]
Finally did enough testing to be confident that this works. Merging..

**Pull request merged**